### PR TITLE
Remove Dimension Entity

### LINKED ISSUE #83: Inline the attributes of the Dimension entity into the DimensionContent
This issue strongly relates to #87.

If we decide to allow only a `stage` and a `locale` attribute, we could also just save this attributes directly on the `DimensionContent` entity and get rid of the `Dimension` completely. 

This would simplify the code of the bundle in a lot of places. For example, in addition to the `DimensionEntity` we could remove the `DimensionRepository`, the `DimensionCollection` and the `DimensionCollectionFactory`.

### PR DESCRIPTION
**Pull request by alexander-schranz**
This contains a few BC breaks which is documented in the. `UPGRADE.md`.

Fixes: https://github.com/sulu/SuluContentBundle/issues/83

# TODO

 - [x] Add missing tests
 - [x] Increase coverage/coveralls to 100%
 - [ ] Create Doctrine Example Migration Class which supports multiple tables

### CODE REVIEW

File: Content/Domain/Model/RoutableTrait.php [hunk 1]
```diff
@@ -15,17 +15,10 @@
...
-    {
-        return $this->getDimension()->getLocale();
-    }
-
     public function getResourceId()
     {
         return $this->getResource()->getId();
     }
 
-    abstract public function getDimension(): DimensionInterface;
```
Comment (niklasnatter) [inline Content/Domain/Model/RoutableTrait.php:28]: 🎉

File: Content/Domain/Repository/DimensionContentRepositoryInterface.php [hunk 2]
```diff
@@ -14,13 +14,15 @@
...
 
 interface DimensionContentRepositoryInterface
 {
+    /**
+     * @param mixed[] $dimensionAttributes
+     */
     public function load(
         ContentRichEntityInterface $contentRichEntity,
-        DimensionCollectionInterface $dimensionCollection
+        array $dimensionAttributes = []
```
Comment (niklasnatter) [inline Content/Domain/Repository/DimensionContentRepositoryInterface.php]: not sure if it makes sense to use `[]` as default value? 🤔

File: Content/Infrastructure/Doctrine/DimensionContentRepository.php [hunk 3]
```diff
@@ -44,36 +45,84 @@ public function __construct(
 
     public function load(
         ContentRichEntityInterface $contentRichEntity,
-        DimensionCollectionInterface $dimensionCollection
+        array $dimensionAttributes = []
```
Comment (niklasnatter) [inline Content/Infrastructure/Doctrine/DimensionContentRepository.php]: not sure if it makes sense to use `[]` as default value? 🤔

File: Content/Infrastructure/Doctrine/MetadataLoader.php [hunk 4]
```diff
@@ -46,12 +45,8 @@ public function loadClassMetadata(LoadClassMetadataEventArgs $event): void
...
 
         if ($reflection->implementsInterface(DimensionContentInterface::class)) {
-            $this->addManyToOne(
-                $event,
-                $metadata,
-                'dimension',
-                DimensionInterface::class
-            );
+            $this->addField($metadata, 'stage', 'string', ['length' => 16, 'nullable' => false]);
+            $this->addField($metadata, 'locale', 'string', ['length' => 5, 'nullable' => true]);
```
Comment (niklasnatter) [inline Content/Infrastructure/Doctrine/MetadataLoader.php]: are you sure that length 5 is enough for the locale? 🙈

File: Content/Domain/Model/ContentRichEntityInterface.php [hunk 5]
```diff
@@ -27,7 +27,7 @@ public function getId();
      */
     public function getDimensionContents(): Collection;
 
-    public function createDimensionContent(DimensionInterface $dimension): DimensionContentInterface;
+    public function createDimensionContent(): DimensionContentInterface;
```
Comment (niklasnatter) [inline Content/Domain/Model/ContentRichEntityInterface.php:30]: shouldnt we pass the `dimensionAttributes` here?

File: Tests/Functional/Content/Infrastructure/Doctrine/DimensionContentRepositoryTest.php [hunk 6]
```diff
@@ -150,9 +115,15 @@ private function createContentRichEntity(): Example
...
     }
 
-    private function createContentDimension(Example $example, DimensionInterface $dimension): ExampleDimensionContent
+    /**
+     * @param mixed[] $dimensionAttributes
+     */
+    private function createContentDimension(Example $example, array $dimensionAttributes = []): ExampleDimensionContent
     {
-        $exampleDimension = new ExampleDimensionContent($example, $dimension);
+        $exampleDimension = new ExampleDimensionContent($example);
```
Comment (niklasnatter) [inline Tests/Functional/Content/Infrastructure/Doctrine/DimensionContentRepositoryTest.php:123]: not sure if we should pass the `dimensionAttributes` directly to the constructor? or if there should be a single setter for them?

File: Tests/Functional/Content/Infrastructure/Sulu/Search/ContentReindexProviderTest.php [hunk 7]
```diff
@@ -16,20 +16,15 @@
...
 use Sulu\Bundle\ContentBundle\Tests\Application\ExampleTestBundle\Entity\ExampleDimensionContent;
-use Sulu\Bundle\ContentBundle\Tests\Functional\BaseTestCase;
 use Sulu\Bundle\ContentBundle\Tests\Traits\CreateExampleTrait;
-use Sulu\Bundle\ContentBundle\Tests\Traits\ModifyExampleTrait;
-use Sulu\Bundle\ContentBundle\Tests\Traits\PublishExampleTrait;
+use Sulu\Bundle\TestBundle\Testing\SuluTestCase;
 use Sulu\Component\HttpKernel\SuluKernel;
 
-class ContentReindexProviderTest extends BaseTestCase
+class ContentReindexProviderTest extends SuluTestCase
```
Comment (niklasnatter) [inline Tests/Functional/Content/Infrastructure/Sulu/Search/ContentReindexProviderTest.php:25]: whats the reason for not extending the `BaseTestCase` here? 🙂

File: Tests/Functional/Content/Infrastructure/Sulu/SmartContent/ContentDataProviderTest.php [hunk 8]
```diff
@@ -73,176 +69,211 @@ public static function setUpBeforeClass(): void
...
+            'de' => [
+                'title' => 'example without categories without tags',
+                'draft' => [
+                    'title' => 'example without categories without tags draft',
+                    'excerptCategories' => [
+                        static::$categoryFoo->getId(),
+                        static::$categoryBar->getId(),
+                    ],
+                    'excerptTags' => [
+                        static::$tagB->getId(),
```
Comment (niklasnatter) [inline Tests/Functional/Content/Infrastructure/Sulu/SmartContent/ContentDataProviderTest.php]: dont we need to use the `name` instead of the `id` for tags?

File: composer.json [hunk 9]
```diff
@@ -86,12 +87,11 @@
         ],
         "test-coverage": "@phpunit --coverage-php Tests/reports/coverage.php --coverage-html Tests/reports/html --log-junit Tests/reports/unit/junit.xml",
         "test-coverage-checker": [
-            "@php vendor/bin/code-coverage-checker \"Tests/reports/coverage.php\" \"line\" \"99.00\" \"Content/Domain\" \"Content/Application\"",
-            "@php vendor/bin/code-coverage-checker \"Tests/reports/coverage.php\" \"line\" \"94.00\" \"Content/Infrastructure\" \"DependencyInjection\""
+            "@php vendor/bin/code-coverage-checker \"Tests/reports/coverage.php\" \"line\" \"100.00\" \"Content\" \"DependencyInjection\""
```
Comment (nnatter) [inline composer.json:90]: 🎉

File: Content/Application/ContentMerger/ContentMerger.php [hunk 10]
```diff
@@ -24,35 +25,47 @@ class ContentMerger implements ContentMergerInterface
...
-        $mergedDimensionContent = $contentRichEntity->createDimensionContent($mostSpecificDimension);
+        $mergedDimensionContent = $contentRichEntity->createDimensionContent();
         $mergedDimensionContent->markAsMerged();
 
         foreach ($dimensionContentCollection as $dimensionContent) {
             foreach ($this->mergers as $merger) {
                 $merger->merge($mergedDimensionContent, $dimensionContent);
             }
+
+            foreach ($dimensionContent::getDefaultAttributes() as $key => $value) {
```
Comment (niklasnatter) [inline Content/Application/ContentMerger/ContentMerger.php]: we could also use the `dimensionAttributes` of the `dimensionContentCollection`?

File: Content/Domain/Model/DimensionContentCollection.php [hunk 11]
```diff
@@ -35,34 +36,55 @@ class DimensionContentCollection implements \IteratorAggregate, DimensionContent
...
-            )->first() ?: null;
-        }
+    public function __construct(
+        array $dimensionContents,
+        array $dimensionAttributes,
+        string $dimensionContentClass
+    ) {
+        $this->dimensionContents = new ArrayCollection($dimensionContents);
+        $this->dimensionContentClass = $dimensionContentClass;
+        $this->defaultDimensionAttributes = \call_user_func([$dimensionContentClass, 'getDefaultAttributes']);
```
Comment (niklasnatter) [inline Content/Domain/Model/DimensionContentCollection.php]: we just used `dimensionContentClass::getDefaultAttributes()` in other places: https://github.com/sulu/SuluContentBundle/blob/107a97625aa4fcdc7de9ab0b782f2edd7146bbd0/Content/Infrastructure/Sulu/Teaser/ContentTeaserProvider.php#L317

File: Content/Domain/Model/DimensionContentCollection.php [hunk 12]
```diff
@@ -35,34 +36,55 @@ class DimensionContentCollection implements \IteratorAggregate, DimensionContent
...
 
-        $localizedDimension = $dimensionCollection->getLocalizedDimension();
-        if ($localizedDimension) {
-            $this->localizedDimensionContent = $dimensionContentArrayCollection->filter(
-                function (DimensionContentInterface $dimensionContent) use ($localizedDimension) {
-                    return $dimensionContent->getDimension()->getId() === $localizedDimension->getId();
-                }
-            )->first() ?: null;
-        }
+        $this->unlocalizedDimensionContent = $this->dimensionContents->filter(
```
Comment (niklasnatter) [inline Content/Domain/Model/DimensionContentCollection.php:70]: not sure if i would set this as private variable, but we can do this in another PR 🙂

File: Content/Domain/Model/DimensionContentCollection.php [hunk 13]
```diff
@@ -75,9 +97,32 @@ public function getLocalizedDimensionContent(): ?DimensionContentInterface
         return $this->localizedDimensionContent;
     }
 
+    public function getDimensionContent(array $dimensionAttributes): ?DimensionContentInterface
+    {
+        $dimensionAttributes = array_merge($this->defaultDimensionAttributes, $dimensionAttributes);
```
Comment (niklasnatter) [inline Content/Domain/Model/DimensionContentCollection.php:102]: maybe just call `dimensionContentClass::getDefaultAttributes()` here instead of saving it on the object?

File: Content/Domain/Model/DimensionContentCollection.php [hunk 14]
```diff
@@ -14,14 +14,15 @@
 namespace Sulu\Bundle\ContentBundle\Content\Domain\Model;
 
 use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Criteria;
 
 /**
  * @implements \IteratorAggregate<DimensionContentInterface>
  */
 class DimensionContentCollection implements \IteratorAggregate, DimensionContentCollectionInterface
```
Comment (niklasnatter) [inline Content/Domain/Model/DimensionContentCollection.php:22]: no a fan of `implements \IteratorAggregate`, but we can talk about that in another PR

File: Content/Domain/Model/DimensionContentCollectionInterface.php [hunk 15]
```diff
@@ -21,4 +21,19 @@ interface DimensionContentCollectionInterface extends \Traversable, \Countable
     public function getUnlocalizedDimensionContent(): ?DimensionContentInterface;
 
     public function getLocalizedDimensionContent(): ?DimensionContentInterface;
+
+    /**
+     * @param mixed[] $dimensionAttributes
+     */
+    public function getDimensionContent(array $dimensionAttributes): ?DimensionContentInterface;
```
Comment (niklasnatter) [inline Content/Domain/Model/DimensionContentCollectionInterface.php:28]: 👍

Comment (niklasnatter) [reply Content/Domain/Model/DimensionContentCollectionInterface.php:28]: maybe we can remove the `getUnlocalizedDimensionContent` and `getLocalizedDimensionContent` in another PR and use this instead

File: Content/Infrastructure/Doctrine/DimensionContentRepository.php [hunk 16]
```diff
@@ -44,36 +45,84 @@ public function __construct(
...
+
+    /**
+     * @param class-string<DimensionContentInterface> $className
+     * @param mixed[] $attributes
+     *
+     * @return mixed[]
+     */
+    private function getAttributes(string $className, array $attributes): array
+    {
+        $defaultValues = \call_user_func([$className, 'getDefaultAttributes']);
```
Comment (niklasnatter) [inline Content/Infrastructure/Doctrine/DimensionContentRepository.php]: we just used `dimensionContentClass::getDefaultAttributes()` in other places: https://github.com/sulu/SuluContentBundle/blob/107a97625aa4fcdc7de9ab0b782f2edd7146bbd0/Content/Infrastructure/Sulu/Teaser/ContentTeaserProvider.php#L317

File: Content/Infrastructure/Doctrine/DimensionContentRepository.php [hunk 17]
```diff
@@ -44,36 +45,84 @@ public function __construct(
...
+        return $criteria;
+    }
+
+    /**
+     * @param class-string<DimensionContentInterface> $className
+     * @param mixed[] $attributes
+     *
+     * @return mixed[]
+     */
+    private function getAttributes(string $className, array $attributes): array
```
Comment (niklasnatter) [inline Content/Infrastructure/Doctrine/DimensionContentRepository.php]: would call this `getEffectiveAttributes`

File: UPGRADE.md [hunk 18]
```diff
@@ -2,6 +2,7 @@
 
 ## 0.5.0
 
+<<<<<<< HEAD
```
Comment (niklasnatter) [inline UPGRADE.md]: merge conflict 🙂

File: UPGRADE.md [hunk 19]
```diff
@@ -68,6 +69,120 @@ parameter. In this case also the `getShowDrafts` was removed from the `ContentTe
...
+-
+        <property name="title" visibility="yes" translation="sulu_admin.title">
+            <field-name>title</field-name>
+            <entity-name>dimensionContent</entity-name>
+
+            <joins ref="dimensionContent"/>
+        </property>
+    </properties>
+</list>
+>>>>>>> 1ac7236... Remove dimension entity
```
Comment (niklasnatter) [inline UPGRADE.md]: merge conflict

File: Tests/Traits/CreateExampleTrait.php [hunk 20]
```diff
@@ -13,51 +13,155 @@
...
-
-        $defaultData = [
-            'template' => $template,
-            'title' => $title,
-            'url' => '/' . (new AsciiSlugger())->slug($title),
-            'article' => '<p>Test article</p>',
-        ];
-
-        $dimensionAttributes = ['locale' => $locale];
+    protected static function createExample(array $dataSet = [], array $options = []): Example
```
Comment (niklasnatter) [inline Tests/Traits/CreateExampleTrait.php:34]: i think i would still use the services of the bundle (eg `ContentPersister`) because it makes things easier to read. but thats not a strong opinion

File: Tests/Traits/CreateExampleTrait.php [hunk 21]
```diff
@@ -13,51 +13,155 @@
...
+                $entityManager->persist($liveUnlocalizedDimension);
+                $createdPublishedUnlocalizedDimension = true;
+            }
+
+            $draft = $data['draft'] ?? null;
+            unset($data['draft']);
+
+            $draftLocalizedDimension = new ExampleDimensionContent($example);
+
+            $setDraftData = function (ExampleDimensionContent $draftLocalizedDimension, array $data) use ($entityManager, $slugger, $locale) {
```
Comment (niklasnatter) [inline Tests/Traits/CreateExampleTrait.php]: this feels quite complex 😕

File: Content/Application/ContentMerger/ContentMerger.php (hunk 10 above)
Comment (alexander-schranz) [reply Content/Application/ContentMerger/ContentMerger.php]: thats a good idea. didn't like this also not todo this inside this loop 👍

File: Content/Domain/Model/ContentRichEntityInterface.php (hunk 5 above)
Comment (alexander-schranz) [reply Content/Domain/Model/ContentRichEntityInterface.php:30]: Don't want that the developer which creates the Content Entity need to implement setting of the `dimensionAttributes` so I removed it here and used a `PropertyAccessor` to set this, which we maybe could replace with setter function (setDimensionAttributes). But don't want that they need to handle this in der entity implementation.

File: Content/Domain/Model/DimensionContentCollection.php (hunk 14 above)
Comment (alexander-schranz) [reply Content/Domain/Model/DimensionContentCollection.php:22]: Sure would currently keep it this way. Maybe you want to create an issue or do we have one?

File: Content/Domain/Model/DimensionContentCollection.php (hunk 12 above)
Comment (alexander-schranz) [reply Content/Domain/Model/DimensionContentCollection.php:70]: Currently the whole logic is build ontop of the `unlocalizedDimensionContent` and `localizedDimensionContent` variable so its not really private. But I think we can refractor this in another PR when we remove that logic from the `MapperInterface`.

File: Content/Domain/Model/DimensionContentCollection.php (hunk 13 above)
Comment (alexander-schranz) [reply Content/Domain/Model/DimensionContentCollection.php:102]: I personally not a big fan of calling external functions which are not part of the codebase multiple times as you don't know what will be there which could have an performance impact. That why I did `cache` it in a private variable inside the collection. But could be changed if you insist:

waltz

File: Content/Domain/Model/DimensionContentCollectionInterface.php (hunk 15 above)
Comment (alexander-schranz) [reply Content/Domain/Model/DimensionContentCollectionInterface.php:28]: Thats also my plan think that would be great to have the collection and the mapper using this function to get the correct dimension.

File: Content/Infrastructure/Doctrine/MetadataLoader.php (hunk 4 above)
Comment (alexander-schranz) [reply Content/Infrastructure/Doctrine/MetadataLoader.php]: Don't want that `locale` is misused again as some kind of region strings and with `en_US` it would be fine but if we want to support `zh_HANS` and `zh_HANT` we would go with `7`. What do you think?

File: Tests/Functional/Content/Infrastructure/Doctrine/DimensionContentRepositoryTest.php (hunk 6 above)
Comment (alexander-schranz) [reply Tests/Functional/Content/Infrastructure/Doctrine/DimensionContentRepositoryTest.php:123]: I would like to not have it in the constructor because then the developer which uses the bundle need to implement the dimension `setters`. A general setter which replaces the current PropertyAccessor service is also fine for me, but not needed from my side.

File: Tests/Functional/Content/Infrastructure/Sulu/Search/ContentReindexProviderTest.php (hunk 7 above)
Comment (alexander-schranz) [reply Tests/Functional/Content/Infrastructure/Sulu/Search/ContentReindexProviderTest.php:25]: oh forgot to remove that, don't wanted to have BaseTestCase think strange to have magic getter which return a service definitions.

File: Tests/Functional/Content/Infrastructure/Sulu/SmartContent/ContentDataProviderTest.php (hunk 8 above)
Comment (alexander-schranz) [reply Tests/Functional/Content/Infrastructure/Sulu/SmartContent/ContentDataProviderTest.php]: No did work with the `ids` in the fixtures for performance reasons.

File: Tests/Traits/CreateExampleTrait.php (hunk 20 above)
Comment (alexander-schranz) [reply Tests/Traits/CreateExampleTrait.php:34]: I would keep it this way, ContentPersister is doing so much stuff, requires a flush for every single entity because it can not generate the url without it. So at current state I would go here with maybe complexity in this fixture class but a better performance in running the tests.

File: Content/Infrastructure/Doctrine/MetadataLoader.php (hunk 4 above)
Comment (alexander-schranz) [reply Content/Infrastructure/Doctrine/MetadataLoader.php]: increased it to 7

File: Tests/Functional/Content/Infrastructure/Sulu/SmartContent/ContentDataProviderTest.php (hunk 8 above)
Comment (alexander-schranz) [reply Tests/Functional/Content/Infrastructure/Sulu/SmartContent/ContentDataProviderTest.php]: refractored the CreateExampleTrait to use specific Mapper services

File: Tests/Traits/CreateExampleTrait.php (hunk 21 above)
Comment (alexander-schranz) [reply Tests/Traits/CreateExampleTrait.php]: refractored to use Mapper Services where possible

File: composer.json [hunk 22]
```diff
@@ -94,6 +95,11 @@
             "@phpunit"
         ],
         "phpunit": "vendor/bin/simple-phpunit",
+        "test-coverage": "@phpunit --coverage-php Tests/reports/coverage.php --coverage-html Tests/reports/html --log-junit Tests/reports/unit/junit.xml --coverage-clover Tests/reports/clover.xml",
+        "test-coverage-checker": [
```
Comment (niklasnatter) [inline composer.json]: think i would call this `check-coverage`

File: composer.json [hunk 23]
```diff
@@ -102,6 +108,7 @@
         "php-cs": "vendor/bin/php-cs-fixer fix --verbose --diff --dry-run",
         "php-cs-fix": "vendor/bin/php-cs-fixer fix",
         "lint-composer": "@composer validate --strict",
+        "lint-schema": "Tests/Application/bin/adminconsole doctrine:schema:validate",
```
Comment (niklasnatter) [inline composer.json]: this is already included in `lint-doctrine`

File: composer.json [hunk 24]
```diff
@@ -94,6 +95,11 @@
             "@phpunit"
         ],
         "phpunit": "vendor/bin/simple-phpunit",
+        "test-coverage": "@phpunit --coverage-php Tests/reports/coverage.php --coverage-html Tests/reports/html --log-junit Tests/reports/unit/junit.xml --coverage-clover Tests/reports/clover.xml",
```
Comment (niklasnatter) [inline composer.json]: think i would call this `test-with-coverage`, `test-coverage`sounds like we are testing the coverage 🙈

File: Content/Domain/Model/ContentRichEntityInterface.php (hunk 5 above)
Comment (niklasnatter) [reply Content/Domain/Model/ContentRichEntityInterface.php:30]: is okay for me 👍

File: Content/Domain/Model/DimensionContentCollection.php (hunk 13 above)
Comment (niklasnatter) [reply Content/Domain/Model/DimensionContentCollection.php:102]: is okay for me 👍

File: Content/Infrastructure/Doctrine/DimensionContentRepository.php [hunk 25]
```diff
@@ -44,36 +45,84 @@ public function __construct(
...
+     * @param class-string<DimensionContentInterface> $className
+     * @param mixed[] $attributes
+     *
+     * @return mixed[]
+     */
+    private function getEffectiveAttributes(string $className, array $attributes): array
+    {
+        $defaultValues = $className::getDefaultAttributes();
+
+        // Ignore any key which is is which has no default values
```
Comment (niklasnatter) [inline Content/Infrastructure/Doctrine/DimensionContentRepository.php]:
```suggestion
        // Ignore keys that are not part of the default attributes
```

File: Tests/Functional/Content/Infrastructure/Sulu/Search/ContentReindexProviderTest.php (hunk 7 above)
Comment (niklasnatter) [reply Tests/Functional/Content/Infrastructure/Sulu/Search/ContentReindexProviderTest.php:25]: is okay for me 👍

File: Tests/Unit/Content/Application/ContentMerger/ContentMergerTest.php [hunk 26]
```diff
@@ -45,40 +45,45 @@ public function testMerge(): void
...
+        $merger1->merge($mergedDimensionContent->reveal(), $dimensionContent1)->shouldBeCalled();
+        $merger2->merge($mergedDimensionContent->reveal(), $dimensionContent1)->shouldBeCalled();
 
-        $merger1->merge($mergedDimensionContent->reveal(), $dimensionContent3->reveal())->shouldBeCalled();
-        $merger2->merge($mergedDimensionContent->reveal(), $dimensionContent3->reveal())->shouldBeCalled();
+        $merger1->merge($mergedDimensionContent->reveal(), $dimensionContent2)->shouldBeCalled();
+        $merger2->merge($mergedDimensionContent->reveal(), $dimensionContent2)->shouldBeCalled();
 
-        $mergedDimensionContent->markAsMerged()->shouldBeCalled();
+        $mergedDimensionContent->setLocale(null) // TODO FIXME find a way to avoid this call (ExampleControllerTest::testPostTriggerUnpublish) currently files without this
```
Comment (niklasnatter) [inline Tests/Unit/Content/Application/ContentMerger/ContentMergerTest.php]:
```suggestion
        $mergedDimensionContent->setLocale(null) // TODO FIXME find a way to avoid this call (ExampleControllerTest::testPostTriggerUnpublish) currently fails without this
```

think the `ExampleControllerTest::testPostTriggerUnpublish` fails because of https://github.com/sulu/SuluContentBundle/issues/145

File: Tests/Unit/Content/Application/ContentWorkflow/ContentWorkflowTest.php [hunk 27]
```diff
@@ -65,19 +61,25 @@ public static function getWorkflowName(): string
             {
                 return 'content_workflow';
             }
+
+            public static function getDefaultAttributes(): array
```
Comment (niklasnatter) [inline Tests/Unit/Content/Application/ContentWorkflow/ContentWorkflowTest.php]: would add this method as default to the `DimensionContentMockWrapperTrait `

File: UPGRADE.md [hunk 28]
```diff
@@ -2,6 +2,120 @@
...
+# Remove dimension relation
+ALTER TABLE test_example_dimension_contents DROP FOREIGN KEY FK_9BFA55B277428AD;
+DROP INDEX IDX_9BFA55B277428AD ON test_example_dimension_contents;
+ALTER TABLE test_example_dimension_contents DROP dimension_id;
+
+# Drop Dimension Table
+DROP TABLE cn_dimensions;
+```
+
+TODO provide here a general doctrine migration which support up/down.
```
Comment (niklasnatter) [inline UPGRADE.md:36]: think we can remove this line. it okay for me to not provide such a script

File: UPGRADE.md [hunk 29]
```diff
@@ -2,6 +2,120 @@
...
++    public function __construct(Example $example)
+     {
+         $this->example = $example;
+-        $this->dimension = $dimension;
+     }
+```
+
+The `DimensionContentInterface` has the `getDimension` removed and will now directly
+need to provide the `getStage`, `setStage`, `getLocale` and `setLocale` methods.
+This is normally done by the traits provided by the ContentBundle.
```
Comment (niklasnatter) [inline UPGRADE.md]:
```suggestion
If you are using the traits provided by the ContentBundle, these methods should be added to your entity automatically.
```

File: UPGRADE.md [hunk 30]
```diff
@@ -2,6 +2,120 @@
...
+
+#### Update your list configuration
+
+If you use the dimension data to be listed in your list you need to change it the following way:
```
Comment (niklasnatter) [inline UPGRADE.md]:
```suggestion
If you use the dimension data in your list configuration, you need to change it the following way:
```

File: Content/Domain/Model/DimensionContentCollection.php (hunk 12 above)
Comment (niklasnatter) [reply Content/Domain/Model/DimensionContentCollection.php:70]: is okay for me 👍

File: Content/Domain/Model/DimensionContentCollectionInterface.php (hunk 15 above)
Comment (niklasnatter) [reply Content/Domain/Model/DimensionContentCollectionInterface.php:28]: will add an issue for this 👍

File: Tests/Traits/CreateExampleTrait.php [hunk 31]
```diff
@@ -13,51 +13,138 @@
...
+        $liveUnlocalizedDimension = null;
+        $createdPublishedUnlocalizedDimension = false;
+
+        foreach ($dataSet as $locale => $data) {
+            $published = $data['published'] ?? false;
+            unset($data['published']);
+
+            // draft data
+            $draft = $data['draft'] ?? null;
+            unset($data['draft']);
```
Comment (niklasnatter) [inline Tests/Traits/CreateExampleTrait.php]: what do you think about using the following array structure:
```php
[
    // unpublished
    'en' => [
        'draft' => [
            'title' => 'draft title',
        ],
    ],
    // published, will fallback to live for draft data
    'de' => [
        'live' => [
            'title' => 'draft and live title',
        ]
    ],
    // published with a different draft
    'fr' => [
        'draft' => [
            'title' => 'draft title',
        ],
        'live' => [
            'title' => 'live title',
        ]
    ],
]
```

i think this would be a bit easier to understand and we would not need a `published` and `draft` key inside of the data array

File: UPGRADE.md [hunk 32]
```diff
@@ -2,6 +2,120 @@
 
 ## 0.5.0
 
+### Dimension Entity was removed
+
+The `Dimension` entity was removed because it had no additional value and did make things
+unnecessary complex.
+
+#### Migrate data into your entity
```
Comment (niklasnatter) [inline UPGRADE.md]:
```suggestion
#### Migrate data into your DimensionContent entity
```

File: UPGRADE.md [hunk 33]
```diff
@@ -2,6 +2,120 @@
...
+DROP INDEX IDX_9BFA55B277428AD ON test_example_dimension_contents;
+ALTER TABLE test_example_dimension_contents DROP dimension_id;
+
+# Drop Dimension Table
+DROP TABLE cn_dimensions;
+```
+
+TODO provide here a general doctrine migration which support up/down.
+
+#### Update your Content Entities
```
Comment (niklasnatter) [inline UPGRADE.md]:
```suggestion
#### Update your ContentRichEntity class and DimensionContent class
```

File: UPGRADE.md [hunk 34]
```diff
@@ -2,6 +2,120 @@
...
+
+In your entity you need to change the createDimensionContent method:
```
Comment (niklasnatter) [inline UPGRADE.md]:
```suggestion
In your "ContentRichEntity" class you need to change the createDimensionContent method:
```

**Review by niklasnatter — COMMENTED** [review]
Looks good! 🙂

File: Tests/Traits/CreateExampleTrait.php (hunk 20 above)
Comment (niklasnatter) [reply Tests/Traits/CreateExampleTrait.php:34]: is okay for me 🙂

File: Content/Domain/Model/DimensionContentInterface.php [hunk 35]
```diff
@@ -15,13 +15,27 @@
...
     public function getResource(): ContentRichEntityInterface;
 
     public function isMerged(): bool;
 
     public function markAsMerged(): void;
+
+    /**
+     * @return mixed[]
+     */
+    public static function getDefaultAttributes(): array;
```
Comment (niklasnatter) [inline Content/Domain/Model/DimensionContentInterface.php]: only noticed this now - should we call this `getDefaultDimensionAttributes`? not sure about it 🙈

**Review by niklasnatter — APPROVED** [review]

**Comment by alexander-schranz** [comment]
🎉
